File: media_server_cpp/src/room_manager.cpp
#include "room_manager.h"
#include <algorithm>
#include <iostream>

namespace onlylang {

bool RoomManager::create_room(const std::string& room_id, const std::string& post_id,
                               const std::string& host_user_id) {
    std::lock_guard<std::mutex> lock(mutex_);

    if (rooms_.find(room_id) != rooms_.end()) {
        return false;
    }

    Room room;
    room.room_id = room_id;
    room.post_id = post_id;  // Changed from classroom_id
    room.host_user_id = host_user_id;
    room.is_active = true;
    room.created_at = std::chrono::steady_clock::now();
    room.last_activity = std::chrono::steady_clock::now();

    rooms_[room_id] = room;
    return true;
}

bool RoomManager::delete_room(const std::string& room_id) {
    std::lock_guard<std::mutex> lock(mutex_);

    auto it = rooms_.find(room_id);
    if (it == rooms_.end()) {
        return false;
    }

    rooms_.erase(it);
    return true;
}

bool RoomManager::room_exists(const std::string& room_id) {
    std::lock_guard<std::mutex> lock(mutex_);
    return rooms_.find(room_id) != rooms_.end();
}

bool RoomManager::add_participant(const std::string& room_id, const std::string& user_id,
                                  const std::string& username, ParticipantRole role,
                                  std::shared_ptr<WebRTCHandler> handler) {
    std::lock_guard<std::mutex> lock(mutex_);

    auto it = rooms_.find(room_id);
    if (it == rooms_.end()) {
        return false;
    }

    Participant participant;
    participant.user_id = user_id;
    participant.username = username;
    participant.role = role;
    participant.webrtc_handler = handler;
    participant.joined_at = std::chrono::steady_clock::now();
    participant.is_active = true;

    it->second.participants[user_id] = participant;
    it->second.last_activity = std::chrono::steady_clock::now();

    return true;
}

bool RoomManager::remove_participant(const std::string& room_id, const std::string& user_id) {
    std::lock_guard<std::mutex> lock(mutex_);

    auto room_it = rooms_.find(room_id);
    if (room_it == rooms_.end()) {
        return false;
    }

    auto participant_it = room_it->second.participants.find(user_id);
    if (participant_it == room_it->second.participants.end()) {
        return false;
    }

    room_it->second.participants.erase(participant_it);
    room_it->second.last_activity = std::chrono::steady_clock::now();

    return true;
}

Room* RoomManager::get_room(const std::string& room_id) {
    std::lock_guard<std::mutex> lock(mutex_);

    auto it = rooms_.find(room_id);
    if (it == rooms_.end()) {
        return nullptr;
    }

    return &it->second;
}

std::vector<std::string> RoomManager::get_active_rooms() {
    std::lock_guard<std::mutex> lock(mutex_);

    std::vector<std::string> active_rooms;
    for (const auto& [room_id, room] : rooms_) {
        if (room.is_active) {
            active_rooms.push_back(room_id);
        }
    }

    return active_rooms;
}

int RoomManager::get_total_rooms() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return static_cast<int>(rooms_.size());
}

int RoomManager::get_total_participants() const {
    std::lock_guard<std::mutex> lock(mutex_);

    int total = 0;
    for (const auto& [room_id, room] : rooms_) {
        total += static_cast<int>(room.participants.size());
    }

    return total;
}

void RoomManager::cleanup_idle_rooms(int timeout_seconds) {
    std::lock_guard<std::mutex> lock(mutex_);

    auto now = std::chrono::steady_clock::now();
    std::vector<std::string> rooms_to_delete;

    for (const auto& [room_id, room] : rooms_) {
        auto idle_time = std::chrono::duration_cast<std::chrono::seconds>(
            now - room.last_activity).count();

        if (idle_time > timeout_seconds) {
            rooms_to_delete.push_back(room_id);
        }
    }

    for (const auto& room_id : rooms_to_delete) {
        rooms_.erase(room_id);
        std::cout << "Cleaned up idle room: " << room_id << std::endl;
    }
}

RoomManager::Stats RoomManager::get_stats() {
    std::lock_guard<std::mutex> lock(mutex_);

    Stats stats;
    stats.total_rooms = static_cast<int>(rooms_.size());
    stats.active_rooms = 0;
    stats.total_participants = 0;
    stats.total_viewers = 0;
    stats.total_hosts = 0;

    for (const auto& [room_id, room] : rooms_) {
        if (room.is_active) {
            stats.active_rooms++;
        }

        for (const auto& [user_id, participant] : room.participants) {
            if (participant.is_active) {
                stats.total_participants++;
                if (participant.role == ParticipantRole::VIEWER) {
                    stats.total_viewers++;
                } else if (participant.role == ParticipantRole::HOST) {
                    stats.total_hosts++;
                }
            }
        }
    }

    return stats;
}

} // namespace onlylang

File: media_server_node/src/RoomManager.ts
/**
 * RoomManager - Manages streaming rooms and peers
 */

import { v4 as uuidv4 } from 'uuid';
import { Room, Peer, RoomStats, ServerStats } from './types';
import { types as mediasoupTypes } from 'mediasoup';
import MediasoupHandler from './MediasoupHandler';

export class RoomManager {
  private rooms: Map<string, Room> = new Map();
  private peers: Map<string, Peer> = new Map();
  private mediasoupHandler: MediasoupHandler;
  private maxRooms: number;
  private maxViewersPerRoom: number;

  constructor(
    mediasoupHandler: MediasoupHandler,
    maxRooms: number,
    maxViewersPerRoom: number
  ) {
    this.mediasoupHandler = mediasoupHandler;
    this.maxRooms = maxRooms;
    this.maxViewersPerRoom = maxViewersPerRoom;
  }

  /**
   * Create a new streaming room
   */
  async createRoom(postId: string, hostUserId: string): Promise<string> {
    if (this.rooms.size >= this.maxRooms) {
      throw new Error(`Maximum rooms (${this.maxRooms}) reached`);
    }

    const roomId = uuidv4();
    const router = await this.mediasoupHandler.createRouter();

    const room: Room = {
      room_id: roomId,
      post_id: postId,
      host_user_id: hostUserId,
      created_at: new Date(),
      is_active: true,
      viewer_count: 0,
      max_viewers: this.maxViewersPerRoom,
      router: router
    };

    this.rooms.set(roomId, room);
    console.log(`✓ Room created: ${roomId} for post ${postId}`);

    return roomId;
  }

  /**
   * Delete a room and disconnect all peers
   */
  async deleteRoom(roomId: string): Promise<boolean> {
    const room = this.rooms.get(roomId);
    if (!room) {
      return false;
    }

    // Remove all peers in this room
    const peersToRemove: string[] = [];
    this.peers.forEach((peer, peerId) => {
      if (peer.room_id === roomId) {
        peersToRemove.push(peerId);
      }
    });

    for (const peerId of peersToRemove) {
      await this.removePeer(peerId);
    }

    // Close router
    if (room.router) {
      room.router.close();
    }

    room.is_active = false;
    this.rooms.delete(roomId);

    console.log(`✓ Room deleted: ${roomId}`);
    return true;
  }

  /**
   * Add a peer (host or viewer) to a room
   */
  async addPeer(
    roomId: string,
    userId: string,
    username: string,
    role: 'host' | 'viewer'
  ): Promise<string> {
    const room = this.rooms.get(roomId);
    if (!room) {
      throw new Error(`Room ${roomId} not found`);
    }

    if (role === 'viewer' && room.viewer_count >= room.max_viewers) {
      throw new Error(`Room ${roomId} is full`);
    }

    const peerId = uuidv4();

    const peer: Peer = {
      peer_id: peerId,
      room_id: roomId,
      user_id: userId,
      username: username,
      role: role,
      is_active: true,
      created_at: new Date(),
      producers: new Map(),
      consumers: new Map()
    };

    this.peers.set(peerId, peer);

    if (role === 'viewer') {
      room.viewer_count++;
    } else if (role === 'host') {
      room.host_peer_id = peerId;
    }

    console.log(`✓ Peer added: ${peerId} (${role}) to room ${roomId}`);

    return peerId;
  }

  /**
   * Remove a peer from a room
   */
  async removePeer(peerId: string): Promise<boolean> {
    const peer = this.peers.get(peerId);
    if (!peer) {
      return false;
    }

    const room = this.rooms.get(peer.room_id);

    // Close producers and consumers
    peer.producers.forEach((producer) => producer.close());
    peer.consumers.forEach((consumer) => consumer.close());

    // Close transports
    if (peer.sendTransport) {
      peer.sendTransport.close();
    }
    if (peer.recvTransport) {
      peer.recvTransport.close();
    }

    // Update room viewer count
    if (room && peer.role === 'viewer') {
      room.viewer_count = Math.max(0, room.viewer_count - 1);
    }

    // If host leaves, close the room
    if (room && peer.role === 'host') {
      await this.deleteRoom(peer.room_id);
    }

    this.peers.delete(peerId);
    console.log(`✓ Peer removed: ${peerId}`);

    return true;
  }

  /**
   * Get a room by ID
   */
  getRoom(roomId: string): Room | undefined {
    return this.rooms.get(roomId);
  }

  /**
   * Get a peer by ID
   */
  getPeer(peerId: string): Peer | undefined {
    return this.peers.get(peerId);
  }

  /**
   * Get statistics for a specific room
   */
  getRoomStats(roomId: string): RoomStats | null {
    const room = this.rooms.get(roomId);
    if (!room) {
      return null;
    }

    return {
      room_id: room.room_id,
      post_id: room.post_id,
      host_user_id: room.host_user_id,
      is_active: room.is_active,
      viewer_count: room.viewer_count,
      created_at: room.created_at.toISOString(),
      bytes_sent: 0, // TODO: Implement async stats collection
      bytes_received: 0 // TODO: Implement async stats collection
    };
  }

  /**
   * Get overall server statistics
   */
  getServerStats(): ServerStats {
    let totalViewers = 0;
    let totalHosts = 0;
    let activeRooms = 0;

    const roomStats: RoomStats[] = [];

    this.rooms.forEach((room) => {
      if (room.is_active) {
        activeRooms++;
        totalViewers += room.viewer_count;
        totalHosts += room.host_peer_id ? 1 : 0;

        roomStats.push({
          room_id: room.room_id,
          post_id: room.post_id,
          host_user_id: room.host_user_id,
          is_active: room.is_active,
          viewer_count: room.viewer_count,
          created_at: room.created_at.toISOString(),
          bytes_sent: 0,
          bytes_received: 0
        });
      }
    });

    return {
      total_rooms: this.rooms.size,
      active_rooms: activeRooms,
      total_peers: this.peers.size,
      total_viewers: totalViewers,
      total_hosts: totalHosts,
      total_bytes_sent: 0,
      total_bytes_received: 0,
      rooms: roomStats
    };
  }

  /**
   * Cleanup idle rooms
   */
  cleanupIdleRooms(idleTimeoutSeconds: number): void {
    const now = Date.now();
    const timeoutMs = idleTimeoutSeconds * 1000;

    const roomsToDelete: string[] = [];

    this.rooms.forEach((room, roomId) => {
      const idleTime = now - room.created_at.getTime();

      // Delete if no host or idle for too long with no viewers
      if (!room.host_peer_id || (room.viewer_count === 0 && idleTime > timeoutMs)) {
        roomsToDelete.push(roomId);
      }
    });

    for (const roomId of roomsToDelete) {
      console.log(`Cleaning up idle room: ${roomId}`);
      this.deleteRoom(roomId);
    }
  }

  /**
   * Get bytes sent for a room from mediasoup stats
   */
  private async getRoomBytesSent(room: Room): Promise<number> {
    let totalBytes = 0;
    try {
      if (room.router) {
        // Aggregate stats from all transports in the room
        this.peers.forEach((peer) => {
          if (peer.room_id === room.room_id) {
            peer.producers.forEach((producer) => {
              // mediasoup stats are async, approximating for now
              totalBytes += 0; // Would need to call producer.getStats()
            });
          }
        });
      }
    } catch (error) {
      console.error('Error getting room bytes sent:', error);
    }
    return totalBytes;
  }

  /**
   * Get bytes received for a room from mediasoup stats
   */
  private async getRoomBytesReceived(room: Room): Promise<number> {
    let totalBytes = 0;
    try {
      if (room.router) {
        this.peers.forEach((peer) => {
          if (peer.room_id === room.room_id) {
            peer.consumers.forEach((consumer) => {
              // Would need to call consumer.getStats()
              totalBytes += 0;
            });
          }
        });
      }
    } catch (error) {
      console.error('Error getting room bytes received:', error);
    }
    return totalBytes;
  }
}

export default RoomManager;
